src/core/or/circuitpadding_machines.c

```diff
@@ -194,29 +131,53 @@ circpad_machine_client_hide_intro_circuits(smartlist_t *machines_sl)
 
   client_machine->name = "client_ip_circ";
 
-  client_machine->conditions.state_mask =
-    CIRCPAD_CIRC_BUILDING|CIRCPAD_CIRC_OPENED;
+  client_machine->conditions.state_mask = CIRCPAD_CIRC_OPENED;
   client_machine->target_hopnum = 2;
 
   /* This is a client machine */
   client_machine->is_origin_side = 1;
 
-  /* For this machine we only pad in client-side introduction circuits even
-   * after introduction has finished */
+  /* We only want to pad introduction circuits, and we want to start padding
+   * only after the INTRODUCE1 cell has been sent, so set the purposes
+   * appropriately.
+   *
+   * In particular we want introduction circuits to blend as much as possible
+   * with general circuits. Most general circuits have the following initial
+   * relay cell sequence (outgoing cells marked in [brackets]):
+   *
+   * [EXTEND2] -> EXTENDED2 -> [EXTEND2] -> EXTENDED2 -> [BEGIN] -> CONNECTED
+   *
+   * followed usually by a [DATA] -> [DATA] -> DATA -> DATA.
+   *
+   * Whereas normal introduction circuits usually look like:
+   *
+   * [EXTEND2] -> EXTENDED2 -> [EXTEND2] -> EXTENDED2 -> [EXTEND2] -> EXTENDED2
+   *  -> [INTRODUCE1] -> INTRODUCE_ACK
+   *
+   * This means that up to the sixth cell, both general and intro circuits have
+   * identical cell sequences. After that we want to mimic the
+   * [DATA] -> [DATA] -> DATA -> DATA sequence, which we achieve by padding
+   * after the INTRODUCE1 has been sent which usually looks like:
+   *
+   *     [INTRODUCE1] -> [PADDING_NEGOTIATE] -> PADDING_NEGOTIATED -> INTRO_ACK
+   *
+   * effectively blending with general circuits up until the end of the circuit
+   * setup. */
   client_machine->conditions.purpose_mask =
-    circpad_circ_purpose_to_mask(CIRCUIT_PURPOSE_C_INTRODUCING)|
     circpad_circ_purpose_to_mask(CIRCUIT_PURPOSE_C_INTRODUCE_ACK_WAIT)|
-    circpad_circ_purpose_to_mask(CIRCUIT_PURPOSE_C_INTRODUCE_ACKED);
+    circpad_circ_purpose_to_mask(CIRCUIT_PURPOSE_C_INTRODUCE_ACKED)|
+    circpad_circ_purpose_to_mask(CIRCUIT_PURPOSE_C_CIRCUIT_PADDING);
 
-  /* XXX Keep the circuit alive even after the introduction has been finished,
+  /* Keep the circuit alive even after the introduction has been finished,
    * otherwise the short-term lifetime of the circuit will blow our cover */
-  // client_machine->manage_circ_lifetime = 1;
+  client_machine->manage_circ_lifetime = 1;
 
-  /* Let the origin-side machine control the lifetime of the circuit. */
+  /* Let the client-side machine control the lifetime of the circuit.
+   * XXX This might act as a fingerprint at the end of the conn. */
```

Comment:
Good point with the XXX. Now that we have manage_circ_lifetime, I don't think we should use should_negotiate_end.. The circuit will stay open until the padding machine shuts down, and then it will time out and get closed by circuit_expire_old_circuits_clientside(). So I would remove this block.

Reply:
Hm, sounds good, but then what would close the relay-side of the circuit? I guess just because the origin-side circuit will close (by expiring), the relay-side of the circuit will also close due to the other end missing? Let me know if this is right and I will remove the block.

Reply:
Yes when the origin circuit is closed, it will cause a DESTROY to get sent towards that middle, which will then close and free its circuit.